### ISSUE TITLE
Handle closed connections gracefully

### ISSUE THREAD

**Issue by elemoine** [peopledoc/procrastinate#255]
On a defer operation, if the db connection obtained from the pool is actually closed (e.g. because of a call to `pg_terminate_backend` on the Postgres side), the defer operation will fail with this error:

```
  File ".../lib/python3.8/site-packages/aiopg/connection.py", line 106, in _ready                                                                        
    state = self._conn.poll()                                                                                                                                                                  
psycopg2.OperationalError: server closed the connection unexpectedly                                                                                                                           
        This probably means the server terminated abnormally                                                                                                                                   
        before or while processing the request.
```

The next defer operation will succeed though, probably because aiopg didn't add the (closed) connection that caused the exception back to the pool.

I am wondering if Procrastinate should detect these cases and raise a specific exception, to give the application an opportunity to handle this as it sees fit.

A similar problem may exist on the worker side as well.

**Comment by elemoine** [comment]
This is more tricky than I thought. I did my testing with `AiopgConnector`, but in "sync" mode. And the behavior is different in "async" mode.

Here are new tests showing this (with the current master branch and the `@wrap_query_exceptions` removed from the `execute_query_one_async` function).

In "sync" mode:

```
$ python
Python 3.8.2 (default, Apr  8 2020, 14:31:25) 
[GCC 9.3.0] on linux
Type "help", "copyright", "credits" or "license" for more information.
>>> from procrastinate_demo import tasks
>>> tasks.sum.defer(a=1, b=2)
22
>>> # server-side disconnect
>>> tasks.sum.defer(a=1, b=2)
Traceback (most recent call last):
  File "/home/elemoine/src/procrastinate/procrastinate/aiopg_connector.py", line 31, in wrapped
  ...
```

Now in "async" mode:

```
$ python -m asyncio
asyncio REPL 3.8.2 (default, Apr  8 2020, 14:31:25) 
[GCC 9.3.0] on linux
Use "await" directly instead of "asyncio.run()".
Type "help", "copyright", "credits" or "license" for more information.
>>> import asyncio
>>> from procrastinate_demo import tasks
>>> await tasks.sum.defer_async(a=1, b=2)
23
>>> # server-side disconnect
>>> await tasks.sum.defer_async(a=1, b=2)
24
>>>
```

Things work in "async" mode. aiopg actually detects that the connection was closed by the server, and provides us with a new connection for the second defer.

It makes some sense actually. In "async" mode we have one long-running event loop. And when the connection is closed by the server aiopg is somehow able to detect it and make the connection as closed (`connection.closed = 2`). In "sync" mode this cannot happen.

Bottom line in that the `@wrap_query_exceptions` decorator that we've added is only useful for the "sync" mode. So I am not sure what we want to do about that.

**Comment by elemoine** [comment]
More information (based on more experiments and studying of the aiopg code):

In the "async" case the closing of the connection (i.e. setting `connection.close` to `2`) happens asynchronously (in the background). And when it happens the connection has no `waiter`, so there's no exception raised. The relevant aiopg code is here: https://github.com/aio-libs/aiopg/blob/master/aiopg/connection.py#L133-L134.

In the "sync" case the closing of the connection happens synchronously – when the connection is used again. This time, when it happens, the connection does have a `waiter`, so an exception is raised.

**Comment by elemoine** [comment]
@ewjoachim with the `@wrap_query_exceptions` decorator in `aiopg_connector.py` being useful only for the "sync" case I am tempted to remove it completely. On the defer side, one should use a `Psycopg2Connector` for the "sync" case. On the worker side, `AiopgConnector` in "async" mode is used. The only thing is deferring from a task. WDYT?

**Comment by ewjoachim** [comment]
No opinion. I think we're on the edge case of an edge case so no strong opinion :)

**Comment by elemoine** [comment]
We actually got an "OperationalError: server closed the connection unexpectedly" error in our aiohttp application this week-end. So it's probably safer to keep `@wrap_query_exceptions` for the moment. I'll just extend its docstrings to provide more context.

**Comment by elemoine** [comment]
This can be closed.

**Comment by hugokitano** [comment]
This thread is from 3 years ago so I would like to get updated advice on this error:

```
% python test_procrastinate.py 2 3                                                                                                                      
App is instantiated in the main Python module (test_procrastinate.py). See https://procrastinate.readthedocs.io/en/stable/discussions.html#top-level-app .
NoneType: None
Scheduling computation of 2 + 3
Future exception was never retrieved
future: <Future finished exception=OperationalError('Connection closed')>
psycopg2.OperationalError: Connection closed
```

What's the best way to handle this? Thanks
